testcases/Lcov.py

```diff
@@ -0,0 +1,126 @@
+
+import logging
+import OpTestLogger
+log = OpTestLogger.optest_logger_glob.get_logger(__name__)
+
+
+class LcovSetup(unittest.TestCase):
+    '''
+    Use the ``ppc64_cpu`` utility to turn SMT on/off and set to each possible
+    SMT mode, which is effectively a CPU hotplug operation.
+    Then start hotplugging CPU cores by using the ``ppc64_cpu`` utility to
+    turn only a specific number of cores on.
+    '''
+
+    def setUp(self):
+        conf = OpTestConfiguration.conf
+        self.cv_HOST = conf.host()
+        self.cv_IPMI = conf.ipmi()
+        self.cv_SYSTEM = conf.system()
+
+    def runTest(self):
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+        self.c = self.cv_SYSTEM.cv_HOST.get_ssh_connection()
+        self.c.run_command("uname -a")
+        res = self.c.run_command("cat /etc/os-release | grep NAME | head -1")
+        if 'SLES' in res[0].strip():
+            self.distro = 'SLES'
+        elif 'Red Hat' in res[0].strip():
+            self.distro = 'RHEL'
+        if self.distro == 'RHEL':
+            cmd = "yum install"
+        elif self.distro == 'SLES':
+            cmd = "zypper install"
+        dep_packages = ['perl*', 'tiny*']
+        for pkg in dep_packages:
+            self.c.run_command(cmd + pkg)
+        time.sleep(5)
+        url = 'git clone https://github.com/linux-test-project/lcov.git'
+        self.c.run_command(cmd)
+        os.chdir('lcov')
+        self.c.run_command("make install")
+
+
+class LcoveCounterZero(unittest.TestCase):
+    '''
+    make the lcov counter to zero before gathering the data
+    '''
+    def setup_test(self):
+        conf = OpTestConfiguration.conf
+        self.cv_SYSTEM = conf.system()
+
+    def runTest(self):
+        '''
+        Running the test
+        '''
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+        self.c = self.cv_SYSTEM.cv_HOST.get_ssh_connection()
+        self.c.run_command("echo 1 > /sys/kernel/debug/gcov/reset")
+        self.c.run_command("lcov --zerocounters")
+
+
+class LcovGatherData(unittest.TestCase):
+    '''
+    collect all gcov data and create an html file
+    '''
+    def setup_test(self):
+        conf = OpTestConfiguration.conf
+        self.cv_SYSTEM = conf.system()
+
+    def runTest(self):
+        '''
+        Running the test
+        '''
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+        self.c = self.cv_SYSTEM.cv_HOST.get_ssh_connection()
+        src_path = '/root/rpmbuild/BUILD/kernel-5.14.0-283.el9/linux-5.14.0-283.el9.ppc64le'
+        gcov_src_path = '/sys/kernel/debug/gcov'
+        gcov_data_path = os.path.join(gcov_src_path, src_path)
+        info_file_path = '/home/test.info'
+        html_file_path = '/home/lcov_results'
+        info_cmd = 'lcov -o' + info_file_path + '-c -f -d' + gcov_data_path + '-b' src_path + '--keep-going' 
+        self.c.run_command(info_cmd)
+        html_cmd = 'genhtml -o' + html_file_path + info_file_path
+        self.c.run_command(info_cmd)
```

Comment:
In general, how will this test be used in an automated flow?
We probably also need a way to build a distro based kernel with GCOV kernel config option enabled.

Reply:
@sacsant For the distro kernel test we thinking to add a new test case like upstream kernel build and calling this as test suite model, sample like 

1-  kernel build with GCov config resue UpstremKernel build Test for distro add new test 
2- zero lcov counter using this test 
3- run host os test case (calling avocado bucket)
4-capture lcov data using test case written here

Reply:
hope we have clarity on this

Reply:
@Naresh-ibm @PraveenPenguin any reason to keep separate test for gcov and lcov,,  i feel both can fit in single test but with different class.. having single test makes it easy to add in CR/CI flow

Reply:
basically this path `src_path = f'/root/kernel-{self.k_version}*/linux-{self.k_version}*'` is only for distro kernel and you need to make it generic to handle upstream source too ?

Reply:
@abdhaleegit as dicussed, for upstream it will be different.
the way it work for distro is different than with upstream. so anything related to upstream will be handled in different script.

Reply:
ok

Reply:
ideally i say GCoveSetup file should InstallDistro Kernel